### PR TITLE
Security update to networking 

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

This PR adds the missing network security to allow handling of certificates and SSLStream
It also updates and fixes various functions of the Networking stack and includes the network integration for the ESP32 target.

The security parts currently only compile for ESP32 as it using Mbedtls and an OpenSSL wrapper. 
The MbedTLS and OpenSSL wrapper are part of the ESP32 IDF.

To implement on STM target will need to add the MbedTLS repo and find another OPenSSL wrapper for it. The best option is to implement the SSL interface directly to MbedTLS and not use the OpenSSL wrapper. Some inital work has been done for this and the build already includes that option.

2 new options for Network security in build.
- NF_SECURITY_OPENSSL
- NF_SECURITY_MBEDTLS

If neither set then you get network without the security code, just uses stubs so any ssl feature should give an error.

### STM32
- get the security working for STM builds. Add mbedtls to build and find openssl wrapper for mbedtls. Or temporarily use the Wolf SSL openssl wrapper until mbedtls interface complete. 
- Test STM32 networking
- Get the F746ZG network working

I am sure there are plenty other things to do 

### ESP32
- Improve wireless support for esp32 
Currently can configure and save SSID / Password for station to connect to AP but doesn't connect until next reboot.  Need to add methods to Join and Disconnect from AP/ status of station
- Currently very limited support for Software AP on ESP32, still needs to be implemented
- Ethernet interface working but commented out as hangs if no ethernet port present 

## How Has This Been Tested?

As part of the this work have been creating a set of network test programs to test individual parts of the network.

These will be committed separately to samples once tidied up and a few more added
but for the moment for a source of test code they are here:- https://github.com/AdrianSoundy/Net-tests

### Tests

- DNS lookup  - OK
- SSLSTream client - OK
- Parse Certificate - OK
- UDP  Send/Receive + unicast - OK
- Network information and events - OK
- TCP Stream connect and Listener - OK
- Set up Wireless connect to AP - OK ( needs more support )
- Multicast test - currently testing
- Mqtt client - OK 
- Desktop echo to UDP messages - OK

Also ported a C# version of MQTT to nanoframework
Its in my github and seems to be working but not much testing done as had issue with VS extension with Esp32 
Branch nanoframework
https://github.com/AdrianSoundy/paho.mqtt.m2mqtt   

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes. (WIll do )
- [ ] All new and existing tests passed.

Signed-off-by: adriansoundy <adriansoundy@gmail.com>

### CODE REVIEW

**Comment by AdrianSoundy** [comment]
Too late to resolve now, early start tomorrow.   Don't why there are conflicts had already merged develop ok

**Comment by josesimoes** [comment]
Suggest that you do a **rebase** from upstream/develop...
No point on trying to fix it at my end and it would require committing to your branch anyway.

**Comment by josesimoes** [comment]
@AdrianSoundy I've committed a few changes with unused parameters otherwise this won't build at AppVeyor because we are now using MinRelSize. The build will fail again because there are still warnings comming from ESPIF and LwIP code.

File: src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp [hunk 1]
```diff
@@ -347,10 +341,11 @@ HRESULT Library_sys_net_native_System_Net_Sockets_NativeSocket::poll___STATIC__B
     }
 
     if(timeout_us < 0)
-        hbTimeout.SetInteger( TIMEOUT_INFINITE);
-    else 
+        hbTimeout.SetInteger( -1 );  // Infinite Timeout
```
Comment (josesimoes) [inline src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp]: shouldn't use _magic_ numbers instead constants or defiines

File: src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp [hunk 2]
```diff
@@ -591,8 +586,8 @@ HRESULT Library_sys_net_native_System_Net_Sockets_NativeSocket::SendRecvHelper(
 
     if(offset + count > arrData->m_numOfElements) NANOCLR_SET_AND_LEAVE(CLR_E_INDEX_OUT_OF_RANGE);    
 
-    hbTimeout.SetInteger( timeout_ms * TIME_CONVERSION__TO_MILLISECONDS);
-        
+// FIXME check working
```
Comment (josesimoes) [inline src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp]: if this is a temporary replacement, the original code should be left commented

File: src/PAL/COM/sockets/ssl/mbedTls/MbedTls.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/mbedTls/MbedTls.cpp]: if we are borrowing code from mbedtls we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/mbedTls/MbedTls.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/mbedTls/MbedTls.h]: if we are borrowing code from mbedtls we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/mbedTls/MbedTls_parse_certificate.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/mbedTls/MbedTls_parse_certificate.cpp]: if we are borrowing code from mbedtls we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/openssl.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/openssl.h]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_accept_internal.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_accept_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_add_cert_auth_internal.cpp [hunk 8]
```diff
@@ -0,0 +1,104 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_add_cert_auth_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_clear_cert_auth_internal.cpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_clear_cert_auth_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_closesocket_internal.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_closesocket_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_connect_internal.cpp [hunk 11]
```diff
@@ -0,0 +1,144 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_connect_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_exit_context_internal.cpp [hunk 12]
```diff
@@ -0,0 +1,34 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_exit_context_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_generic_init_internal.cpp [hunk 13]
```diff
@@ -0,0 +1,222 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_generic_init_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_initialize_internal.cpp [hunk 14]
```diff
@@ -0,0 +1,27 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_initialize_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_parse_certificate_internal.cpp [hunk 15]
```diff
@@ -0,0 +1,240 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_parse_certificate_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_pending_internal.cpp [hunk 16]
```diff
@@ -0,0 +1,30 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_pending_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_read_internal.cpp [hunk 17]
```diff
@@ -0,0 +1,47 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_read_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_types.cpp [hunk 18]
```diff
@@ -0,0 +1,225 @@
+
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_types.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_uninitialize_internal.cpp [hunk 19]
```diff
@@ -0,0 +1,44 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_uninitialize_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/COM/sockets/ssl/openssl/ssl_write_internal.cpp [hunk 20]
```diff
@@ -0,0 +1,50 @@
+//
```
Comment (josesimoes) [inline src/PAL/COM/sockets/ssl/openssl/ssl_write_internal.cpp]: if we are borrowing code from openssl we need to add mention to this in the license header

File: src/PAL/Lwip/lwIP_Sockets.cpp [hunk 21]
```diff
@@ -4,8 +4,21 @@
 // See LICENSE file in the project root for full license information.
```
Comment (josesimoes) [inline src/PAL/Lwip/lwIP_Sockets.cpp]: if we are borrowing code from lwIP we need to add mention to this in the license header

File: targets/FreeRTOS/ESP32_DevKitC/CMakeLists.txt [hunk 22]
```diff
@@ -203,10 +170,25 @@ target_include_directories(${NANOCLR_PROJECT_NAME}.elf PUBLIC
...
+    set (EXTRA_LIBS ${EXTRA_LIBS} NetworkLib)
+    target_link_libraries(${NANOCLR_PROJECT_NAME}.elf ${EXTRA_LIBS})
+    SET( CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} ${GCC_COVERAGE_COMPILE_FLAGS}" )
+
+    # Inhibit warnings so we can compile 3rd party code (Lwip)
+    set_target_properties(NetworkLib PROPERTIES COMPILE_FLAGS " -w " )
+endif()
 
-# set extra compiler flags, some are related with the build type
+# set platform for NanoClr
```
Comment (josesimoes) [inline targets/FreeRTOS/ESP32_DevKitC/CMakeLists.txt]: correct casing is nanoCLR

File: targets/FreeRTOS/ESP32_DevKitC/Network/targetHAL_Network.cpp [hunk 23]
```diff
@@ -11,10 +11,6 @@
 
 extern "C" void set_signal_sock_function( void (*funcPtr)() );
 
-
-// NVS parameters for Ethernet config
```
Comment (josesimoes) [inline targets/FreeRTOS/ESP32_DevKitC/Network/targetHAL_Network.cpp]: sure that these aren't needed anymore?

**Comment by AdrianSoundy** [comment]
ok thanks. I'll have a look. The nanoframework one is working but after a
while if gets an out of range exception.

File: src/PAL/COM/sockets/ssl/mbedTls/MbedTls.cpp (hunk 3 above)
Comment (AdrianSoundy) [reply src/PAL/COM/sockets/ssl/mbedTls/MbedTls.cpp]: not borrowing code from, just calling  mbedTLS

File: src/PAL/COM/sockets/ssl/openssl/ssl_uninitialize_internal.cpp (hunk 19 above)
Comment (AdrianSoundy) [reply src/PAL/COM/sockets/ssl/openssl/ssl_uninitialize_internal.cpp]: All the SSL and ssl_xxxx_xxx are original part of .NetMF. We are just calling the openssl wrapper functions around mbedTLS

File: src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp (hunk 1 above)
Comment (AdrianSoundy) [reply src/HAL/System.Net/sys_net_native_System_Net_Sockets_NativeSocket.cpp]: Unfortunately the TIMEOUT_INFINITE doesn't work it just ends up as a small timeout. 
-1 is a special case in CLR_RT_ExecutionEngine::InitTimeout() which also doesn't use a constant/define.

**Pull request merged**